src/Expression/ExpressionBuilder.php

```diff
@@ -17,9 +22,99 @@
  */
 class ExpressionBuilder implements ExpressionBuilderInterface
 {
+    public function __construct(private QueryBuilderInterface|null $queryBuilder = null)
+    {
+    }
+
     public function build(Expression $expression, array &$params = []): string
     {
-        $params = array_merge($params, $expression->getParams());
-        return $expression->__toString();
+        $sql = $expression->__toString();
+        $expressionParams = $expression->getParams();
+
+        if (empty($expressionParams)) {
+            return $sql;
+        }
+
+        if ($this->queryBuilder === null || isset($expressionParams[0])) {
+            $params = array_merge($params, $expressionParams);
+            return $sql;
+        }
+
+        $sql = $this->appendParams($sql, $expressionParams, $params);
+
+        return $this->replaceParamExpressions($sql, $expressionParams, $params);
+    }
+
+    private function appendParams(string $sql, array &$expressionParams, array &$params): string
+    {
+        $nonUniqueParams = array_intersect_key($expressionParams, $params);
+        $params += $expressionParams;
+
+        if (empty($nonUniqueParams)) {
+            return $sql;
+        }
+
+        $patterns = [];
+        $replacements = [];
+
+        /** @var string $name */
```

Comment:
`$name` can be integer, params may be list (int keys) or array with string keys.

Reply:
It is not possible due to `return` on line 40 above
https://github.com/yiisoft/db/pull/806/files#diff-11971291de3a5da6bd65a674be0a96555c044e0dbfa49c18a4a42554edfddf92R40